### ISSUE TITLE
Some colors are awful and barely readable in emacsclient (fix suggested)

### ISSUE THREAD

**Issue by memeplex** [hlissner/emacs-doom-themes#210]
See some screenshots for magit using tomorrow night:

![image](https://user-images.githubusercontent.com/2845433/43290406-667447fe-9104-11e8-8668-042833a5b315.png)

![image](https://user-images.githubusercontent.com/2845433/43290427-70aef214-9104-11e8-8310-5f1c46812fbc.png)

Weird enough it only happens in client-server mode. Launching emacs --daemon and then emacsclient -c makes the non-highlighted colors look like in the screenshot above. If I simply run emacs they look fine.

**Comment by memeplex** [comment]
Mmmmh, now I restarted my emacs daemon and everything look much better, don't know what could have happened since I never touched any face during the session. But I did install magit in the middle of the session, maybe it overwrote some of the faces the first time it ran or something... really weird stuff. I'm going to close this and reopen if I see it happening again. Sorry.

**Comment by memeplex** [comment]
I'm now able to reproduce it. Weird enough it only happens in client-server mode. Launching `emacs --daemon` and then `emacsclient -c` makes the non-highlighted colors look like in the screenshot above. If I simply run `emacs` they look fine. Any idea why this could be happening?

**Comment by memeplex** [comment]
Please read the follow up at https://github.com/magit/magit/issues/3526 which shed some light on the issue and suggest a couple of workarounds. Also this PR that fixes a similar issue for spacemacs https://github.com/nashamri/spacemacs-theme/pull/89/files. Notice that most colors still are fine in client/server mode except for the ones computed by darken or blend. If the server assumes being in a terminal and these computed colors are not in that terminal color range, maybe emacs could be tricked into believing it's a 24bit terminal, as a third workaround that I haven't try out.

**Comment by memeplex** [comment]
I've been doing some debugging and the detailed results are given in https://github.com/magit/magit/issues/3526#issuecomment-408563976.

TL;DR the problem is that `doom-name-to-rgb` uses `color-values` which depends on the current type of frame or terminal. I think that, except for symbolic color names, the usage of `color-values` is completely avoidable. Avoiding it will result in more robust doom themes that don't require reloading them on each new frame. In other words, transforming a rgb string to a triplet shouldn't depend on the current display. The face color should be set to the exact color and the mapping to a display color should carried on by emacs display abstraction layer as the last step. By calling `color-values` you're prematurely setting a device supported color which is not the exact color more capable devices (for instance, X frames) would have set.

**Comment by memeplex** [comment]
Suggestion, use this as an alternative to `color-values`:

```
(defun doom-color-values (color &optional frame)
  (if (not (string-prefix-p "#" color))
      (color-values color frame)
    (mapcar (lambda (s) (string-to-number s 16))
            (list (substring color 1 3)
                  (substring color 3 5)
                  (substring color 5 7)))))
```

And try to refrain from using symbolic color names when color transformations (blend) are involved. This way color computation will be always independent of the frame.

**Comment by memeplex** [comment]
Digging into the implementation of `color-values` I've found a better alternative: instead of writing a display agnostic version of `color-values` you can use the inbuilt `tty-color-standard-values` which is itself used by `tty-color-values` but returns the exact color that is later approximated according to the display capabilities:

```
(tty-color-standard-values COLOR)

Return standard RGB values of the color COLOR.

The result is a list of integer RGB values--(RED GREEN BLUE).
These values range from 0 to 65535; white is (65535 65535 65535).

The returned value reflects the standard X definition of COLOR,
regardless of whether the terminal can display it, so the return value
should be the same regardless of what display is being used.
```

It's pretty similar to the function I suggested above but it also:

1. Supports `rgb:...` style color specifications
2. Supports symbolic colors using `color-name-rgb-alist`
3. Is already written and tested!

For example, when `TERM=xterm`:

```
(tty-color-standard-values "#b294bb")
  ---> (45568 37888 47872)

(color-values "#b294bb")
  ---> (52685 0 52685)
```

The first triplet is the same one I would have got in a X frame.

**Comment by hlissner** [comment]
Thank you so much for looking into this! It's an issue I've been vaguely aware of for a while, but haven't had much time to look into. Your prognosis makes perfect sense, and your highly detailed report couldn't make it any easier for me! I'll experiment with your suggestions tonight or tomorrow and report back.

**Comment by hlissner** [comment]
After some testing, this certainly does work better. I think we'll go with getting rid of the frame parameter. Would you like to file a PR or shall I do it?

**Comment by memeplex** [comment]
Do it as you think is faster and easier for you! If you prefer me to submit
a PR I have no problem.

**Comment by hlissner** [comment]
As of 1cdc201, this should be resolved. Feel free to reopen this if that isn't the case.

Thanks again for all your help and hard work!

**Comment by memeplex** [comment]
Great, thanks!
